File: js/examples/test-system-endpoint.js
require('dotenv').config();
const OptikpiDataPipelineSDK = require('@optikpi/datapipeline-sdk').default;
const { SystemEvent } = require('@optikpi/datapipeline-sdk')

/**
 * Example of how to send system events using the SDK
 */

// Configuration - Read from environment variables
const API_BASE_URL = process.env.API_BASE_URL;
const AUTH_TOKEN = process.env.AUTH_TOKEN;
const ACCOUNT_ID = process.env.ACCOUNT_ID;
const WORKSPACE_ID = process.env.WORKSPACE_ID;

// Validate required environment variables
if (!AUTH_TOKEN || !ACCOUNT_ID || !WORKSPACE_ID) {
  console.error('❌ Error: Missing required environment variables!');
  console.error('   Please set: AUTH_TOKEN, ACCOUNT_ID, WORKSPACE_ID');
  process.exit(1);
}

// Initialize SDK
const sdk = new OptikpiDataPipelineSDK({
  authToken: AUTH_TOKEN,
  accountId: ACCOUNT_ID,
  workspaceId: WORKSPACE_ID,
  baseURL: API_BASE_URL
});

// System event data - event_data as object
const SYSTEM_EVENT_OBJECT = new SystemEvent({
  account_id: ACCOUNT_ID,
  workspace_id: WORKSPACE_ID,
  event_category: 'SystemEvent',
  event_name: 'CampaignTrigger',
  event_id: `evt_sys_${Date.now()}`,
  event_time: new Date().toISOString(),
  event_data: {
    campaign_id: 'camp_001',
    action: 'start',
    segment: 'vip'
  }
});

// System event data - event_data as JSON string
const SYSTEM_EVENT_STRING = new SystemEvent({
  account_id: ACCOUNT_ID,
  workspace_id: WORKSPACE_ID,
  event_category: 'SystemEvent',
  event_name: 'ManualAction',
  event_id: `evt_sys_${Date.now() + 1}`,
  event_time: new Date().toISOString(),
  event_data: JSON.stringify({
    action: "notify",
    target: "user_list_1",
    payload: {}
  })
});

// Test system endpoint
async function testSystemEndpoint() {
  try {
    console.log('🚀 Testing System (Back Office) Endpoint');
    console.log('========================================');
    console.log('Configuration:');
    console.log(`API Base URL: ${API_BASE_URL}`);
    console.log(`Account ID: ${ACCOUNT_ID}`);
    console.log(`Workspace ID: ${WORKSPACE_ID}`);

    // Test 1: event_data as object
    console.log('\n📋 Test 1: event_data as object');
    console.log('Payload:', JSON.stringify(SYSTEM_EVENT_OBJECT, null, 2));

    // Validate
    const val1 = SYSTEM_EVENT_OBJECT.validate();
    if (!val1.isValid) {
      console.error('❌ Test 1 Validation failed:', val1.errors);
      process.exit(1);
    }
    console.log('✅ System event validated successfully!');

    const startTime1 = Date.now();
    const result1 = await sdk.sendSystemEvent(SYSTEM_EVENT_OBJECT);
    const endTime1 = Date.now();

    if (result1.success) {
      console.log('\n✅ Test 1 Success!');
      console.log(`HTTP Status: ${result1.status}`);
      console.log(`Response Time: ${endTime1 - startTime1}ms`);
      console.log('Response:', JSON.stringify(result1.data, null, 2));
    } else {
      console.log('\n❌ Test 1 Failed!');
      console.log(`HTTP Status: ${result1.status || 'N/A'}`);
      console.log('Error:', result1.error);
      console.log('Data:', JSON.stringify(result1.data, null, 2));
    }

    // Wait 1.5s
    await new Promise(resolve => setTimeout(resolve, 1500));

    // Test 2: event_data as JSON string
    console.log('\n📋 Test 2: event_data as JSON string');
    console.log('Payload:', JSON.stringify(SYSTEM_EVENT_STRING, null, 2));

    // Validate
    const val2 = SYSTEM_EVENT_STRING.validate();
    if (!val2.isValid) {
      console.error('❌ Test 2 Validation failed:', val2.errors);
      process.exit(1);
    }
    console.log('✅ System event validated successfully!');

    const startTime2 = Date.now();
    const result2 = await sdk.sendSystemEvent(SYSTEM_EVENT_STRING);
    const endTime2 = Date.now();

    if (result2.success) {
      console.log('\n✅ Test 2 Success!');
      console.log(`HTTP Status: ${result2.status}`);
      console.log(`Response Time: ${endTime2 - startTime2}ms`);
      console.log('Response:', JSON.stringify(result2.data, null, 2));
    } else {
      console.log('\n❌ Test 2 Failed!');
      console.log(`HTTP Status: ${result2.status || 'N/A'}`);
      console.log('Error:', result2.error);
      console.log('Data:', JSON.stringify(result2.data, null, 2));
    }

    console.log('\n📊 Summary');
    console.log('==========');
    console.log(`Test 1 (object): ${result1.success ? '✅ PASSED' : '❌ FAILED'}`);
    console.log(`Test 2 (string):  ${result2.success ? '✅ PASSED' : '❌ FAILED'}`);
  } catch (error) {
    console.error('\n❌ Error occurred!');
    console.error('==================');
    console.error('Error:', error.message);
  }
}

if (require.main === module) {
  testSystemEndpoint();
}

module.exports = {
  testSystemEndpoint,
  SYSTEM_EVENT_OBJECT,
  SYSTEM_EVENT_STRING,
  sdk
};

File: java/examples/src/main/java/com/optikpi/examples/TestSystemEndpoint.java
package com.optikpi.examples;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

import com.optikpi.datapipeline.ClientConfig;
import com.optikpi.datapipeline.OptikpiDataPipelineSDK;
import com.optikpi.datapipeline.model.SystemEvent;
import com.optikpi.datapipeline.model.ValidationResult;

import io.github.cdimascio.dotenv.Dotenv;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;

/**
 * Example of how to send system events using the SDK
 */
public class TestSystemEndpoint {

    private static final ObjectMapper mapper = new ObjectMapper()
            .enable(SerializationFeature.INDENT_OUTPUT);

    public static void main(String[] args) {
        Dotenv dotenv = Dotenv.configure()
                .directory(".")
                .filename(".env")
                .load();

        String authToken = dotenv.get("AUTH_TOKEN");
        String accountId = dotenv.get("ACCOUNT_ID");
        String workspaceId = dotenv.get("WORKSPACE_ID");
        String baseUrl = dotenv.get("API_BASE_URL");

        if (authToken == null || accountId == null || workspaceId == null) {
            System.err.println("Error: Missing required environment variables");
            System.exit(1);
        }

        ClientConfig config = new ClientConfig(authToken, accountId, workspaceId);
        config.setBaseUrl(baseUrl);

        OptikpiDataPipelineSDK sdk = new OptikpiDataPipelineSDK(config);

        System.out.println("🚀 Testing System (Back Office) Endpoint");
        System.out.println("========================================");
        System.out.println("Configuration:");
        System.out.println("📌 API Base URL: " +  config.getBaseUrl());
        System.out.println("👤 Account ID: " + config.getAccountId());
        System.out.println("🏢 Workspace ID: " + config.getWorkspaceId());
        System.out.println();    
        
        // Test 1: event_data as object (Map)
        System.out.println("📋 Test 1: event_data as object");
        testSystemEventWithObject(sdk, accountId, workspaceId);

        System.out.println("\n------------------------------------------------\n");

        // Test 2: event_data as JSON string
        System.out.println("📋 Test 2: event_data as JSON string");
        testSystemEventWithJsonString(sdk, accountId, workspaceId);
    }

    private static void testSystemEventWithObject(OptikpiDataPipelineSDK sdk, String accountId, String workspaceId) {
        try {
            SystemEvent event = new SystemEvent();
            event.setAccountId(accountId);
            event.setWorkspaceId(workspaceId);
            event.setEventCategory("SystemEvent");
            event.setEventName("Campaign Trigger");
            event.setEventId("evt_sys_" + System.currentTimeMillis());
            event.setEventTime(Instant.now().toString());
            
            Map<String, Object> eventData = new HashMap<>();
            eventData.put("campaign_id", "camp_001");
            eventData.put("action", "start");
            eventData.put("segment", "vip");
            event.setEventData(eventData);

            ValidationResult valid = event.validate();
            if (!valid.isValid()) {
                System.out.println("❌ Validation Failed!");
                System.out.println("Errors: " + valid.getErrors());
                return;
            }
            
            System.out.println("✅ System event object validated successfully!");
            executeRequest(sdk, event);
        } catch (Exception e) {
            System.err.println("❌ Object test failed: " + e.getMessage());
        }
    }

    private static void testSystemEventWithJsonString(OptikpiDataPipelineSDK sdk, String accountId, String workspaceId) {
        try {
            SystemEvent event = new SystemEvent();
            event.setAccountId(accountId);
            event.setWorkspaceId(workspaceId);
            event.setEventCategory("SystemEvent");
            event.setEventName("Manual Action");
            event.setEventId("evt_sys_" + (System.currentTimeMillis() + 1));
            event.setEventTime(Instant.now().toString());
            
            // event_data as JSON string
            String jsonString = "{\"action\":\"notify\",\"target\":\"user_list_1\",\"payload\":{}}";
            event.setEventData(jsonString);

            ValidationResult valid = event.validate();
            if (!valid.isValid()) {
                System.out.println("❌ Validation Failed!");
                System.out.println("Errors: " + valid.getErrors());
                return;
            }

            System.out.println("✅ System event with JSON string validated successfully!");
            executeRequest(sdk, event);
        } catch (Exception e) {
            System.err.println("❌ JSON string test failed: " + e.getMessage());
        }
    }

    private static void executeRequest(OptikpiDataPipelineSDK sdk, SystemEvent event) {
        try {
            System.out.println("📋 Payload:");
            System.out.println(mapper.writeValueAsString(event));

            long start = System.currentTimeMillis();
            var response = sdk.sendSystemEvent(event);
            long end = System.currentTimeMillis();

            System.out.println("⏱ Response Time: " + (end - start) + "ms");
            System.out.println("HTTP Status: " + response.getStatus());

            if (response.isSuccess()) {
                System.out.println("✅ SUCCESS!");
                System.out.println("Response: " + mapper.writeValueAsString(response.getData()));
            } else {
                System.out.println("❌ FAILED!");
                System.out.println("Error: " + response.getError());
            }
        } catch (Exception e) {
            System.err.println("❌ Request execution failed: " + e.getMessage());
        }
    }
}
